### ISSUE TITLE
Support for native i18n / $localize + localized markdown

### ISSUE THREAD

**Issue by michaelkrog** [ng-doc/ng-doc#173]
### Description

Support for Angular's native i18n could really be awesome - especially when combined with  a locale-aware markdown retriever. Coming from EU, documenting products in multiple languages is a given as there is no international language covering the area. The same goes for many other parts of the world.

### Proposed solution

I propose 2 elements to support i18n.

### 1) Support native i18n
Supporting Angular's native i18n goes a long way when handling texts in ts-files (fx. titles).
```
const InstallationPage: NgDocPage = {
  title: $localize`:@@InstallationPage.Title:Installation`,
  mdFile: './index.md'
};
```

### 2) Add ability to resolve markdown-file based on locale
The simplest way to translate markdown-files is to have a markdown-file for each language.  The mdFile property could allow for defining the markdown file per language like this:

```
const InstallationPage: NgDocPage = {
  title: $localize`:@@InstallationPage.Title:Installation`,
  mdFile: {
    'sourceLocale': './index.md',
    'locales': {
      'da': './index.da.md'
    } 
  }
};
```

Then to keep config to a minimum, default behavior for mdFile if not specified, could be that it looks for `./index.<locale>.md` with a fallback to `./index.md`.

That way the config could be minified to this:
```
const InstallationPage: NgDocPage = {
  title: $localize`:@@InstallationPage.Title:Installation`,
};
```

### Alternatives considered

The only alternative I currently know of is to create duplicate projects for each language.

**Comment by skoropadas** [comment]
Thank you for the ticket! Yes, this is quite a complex topic, I have returned to it several times and thought about how it could be implemented. What you are suggesting was one of my first thoughts, but the main problem lies in the builder.

If we decide to create a separate `.md` file for each language, it means that you will have to maintain several identical `.md` files, the main difference being the language. This is very similar to creating and maintaining separate applications for each language. This solution will create additional load on the builder and increase the build time even for local development because NgDoc will have to rebuild all languages for the current page, which may not be very pleasant.

Therefore, I thought that creating separate applications for each language is generally a good solution, they could share some basic components and settings, but have different `angular.json` files and `.md` files. This way, you would not be affected by the number of languages, and the build process would be equally fast for only one running application.

I generally like the idea of a separate application, it requires a little more actions, but it is quite flexible. I think we could go a little further with it to share not only components between applications but also `.md` and `ng-doc.page.ts` files, and add some kind of API so you could use keys instead of text, for example:

```twig
# {{ 'InstallationTitle' | translate }}

{{ 'InstallationBody' | translate }}

<some-code-block/>

{{ 'InstallationCodeDescription' | translate }}

# {{ 'InstallationDemoTitle' | translate }}
{{ 'InstallationDemoBody' | translate }}

{{ NgDocAction.demo("MyDemoComponent") }}
```

This way you could create a separate application for each new language, which you can configure as you wish, but which will have its `angular.json` and `ng-doc.config.ts` files + a file with translation for a specific language, while reusing pages. In this case, you would not need to maintain a demos or playgrounds for each language separately, and by making changes once, your applications for all languages would be updated. But of course, this idea also requires careful consideration.

**Comment by michaelkrog** [comment]
Thank you for such a thorough explanation.
I understand the problem better now.

A few cents from me..

> If we decide to create a separate .md file for each language, it means that you will have to maintain several identical .md files, the main difference being the language. This is very similar to creating and maintaining separate applications

I agree to some extend. But the main difference as I see it is that it is very clear what has been translated and what has not, when all translated md-files for a page is in the same folder.

>  This solution will create additional load on the builder and increase the build time even for local development 

I understand. That is a problem.
For development though, what Angular normally does is serving just one locale. Only production builds generates alle locales.

> This way you could create a separate application for each new language, which you can configure as you wish

That's nice. Flexibility is great. But so is recognizability.

When an Angular developer that is used to Angular's native i18n comes to ng-doc, the easiest path for him would be that i18n works the same as with other Angular apps, namely: One app generates a build per locale.


## Can I take a go at it?
I tried to add @angular/localize to an ng-doc project, but it didn't pickup the $localize values in the code. 

If I were to take a crack at it just to see how it works out, can you perhaps guide me to:
* Why does @angular/localize (ng extract-i18n) not pick up the $localize entries in the code? 🤔
* Where should I start if I wanted to add logic for selecting which md-file to render for a page?

**Comment by skoropadas** [comment]
> I agree to some extend. But the main difference as I see it is that it is very clear what has been translated and what has not, when all translated md-files for a page is in the same folder.

I think I could implement it as well. For example, you could have a `shared` folder with all ng-doc pages that use special syntax as I showed above, and applications pass the translation through the `ng-doc.config.ts` file, where you could also specify the default language that ng-doc could use if something is not translated. I would stick more to this solution to relieve users from supporting multiple identical .md files.

> I understand. That is a problem. For development though, what Angular normally does is serving just one locale. Only production builds generates alle locales.

Yes, but the main problem is that ng-doc doesn't know anything about which page is open in the browser or how the dev-server is launched It is hidden inside the Angular builder. However, in theory, this could be configured by specifying different `ng-doc.config.ts` for build and serve targets.

> That's nice. Flexibility is great. But so is recognizability.
> 
> When an Angular developer that is used to Angular's native i18n comes to ng-doc, the easiest path for him would be that i18n works the same as with other Angular apps, namely: One app generates a build per locale.

I understand, but I'm not sure that this will be implemented using Angular's native i18n. It all depends on how the pages are created for your application. The content you write in a markdown file is not passed as a template for the Angular component of your page but is rendered using `innerHTML`. This was also discussed in #163. I'm going to add experimental support for rendering content using component templates, which should likely make i18n partially work.

But it may also not work. The point is that ng-doc not only converts .md content to html but also performs post-processing on the builder side for each of your pages, making the html interactive, replacing keywords with links, and also compiling indexes for search etc.

In general, I would say that indexes and keywords are the two main reasons why native i18n would not work in ng-doc, the builder performs these steps to finally generate all Angular components and only then start the Angular builder to build the application as usual. Therefore, for translating the documentation, something custom will most likely be used.


> * Why does @angular/localize (ng extract-i18n) not pick up the $localize entries in the code? 🤔

I think I answered that above :)

> * Where should I start if I wanted to add logic for selecting which md-file to render for a page?

At the moment, this is not possible, and I think it wouldn't help you because you also have a `title` field that most likely needs to be translated too. Therefore, the only solution at the moment is to create a second application where you reuse `main.ts`, `app.component.ts`, and other unchanged components, and create separate files for pages and categories, which will be duplicated for each of the languages but will already have translated content inside.

**Comment by eurusik** [comment]
Yes, it would be great to add the ability to use a single file.md from the shared folder in different apps, as well as support for translate_KEY to which various translations can be attached!

**Comment by ousc** [comment]
> Thank you for the ticket! Yes, this is quite a complex topic, I have returned to it several times and thought about how it could be implemented. What you are suggesting was one of my first thoughts, but the main problem lies in the builder.
> 
> If we decide to create a separate `.md` file for each language, it means that you will have to maintain several identical `.md` files, the main difference being the language. This is very similar to creating and maintaining separate applications for each language. This solution will create additional load on the builder and increase the build time even for local development because NgDoc will have to rebuild all languages for the current page, which may not be very pleasant.
> 
> Therefore, I thought that creating separate applications for each language is generally a good solution, they could share some basic components and settings, but have different `angular.json` files and `.md` files. This way, you would not be affected by the number of languages, and the build process would be equally fast for only one running application.
> 
> I generally like the idea of a separate application, it requires a little more actions, but it is quite flexible. I think we could go a little further with it to share not only components between applications but also `.md` and `ng-doc.page.ts` files, and add some kind of API so you could use keys instead of text, for example:
> 
> ```
> # {{ 'InstallationTitle' | translate }}
> 
> {{ 'InstallationBody' | translate }}
> 
> <some-code-block/>
> 
> {{ 'InstallationCodeDescription' | translate }}
> 
> # {{ 'InstallationDemoTitle' | translate }}
> {{ 'InstallationDemoBody' | translate }}
> 
> {{ NgDocAction.demo("MyDemoComponent") }}
> ```
> 
> This way you could create a separate application for each new language, which you can configure as you wish, but which will have its `angular.json` and `ng-doc.config.ts` files + a file with translation for a specific language, while reusing pages. In this case, you would not need to maintain a demos or playgrounds for each language separately, and by making changes once, your applications for all languages would be updated. But of course, this idea also requires careful consideration.

For documents that need to be updated frequently and have a lot of content, using the key method may be a disaster. When writing documents, more splitting work needs to be done. The documents will be broken up and jump in multiple files.
Unlike the i18n solution of the website, such as the `transloco` I am using, I think it is better to directly maintain multiple markdowns for documents, which is easier to write and manage. The various text contents of the website are usually scattered and scattered in various divs. The document is holistic and coherent.
Of course, it would be better to support such pipeline operations as a supplement.
And I think that for ng-doc, there is indeed a need for an i18n solution at present, in addition to building multiple programs. I think the first solution proposed under this question is a feasible solution. The only disadvantage is that the builder needs to build for more markdown files. But I think this is not unacceptable.
Thank you very much for reading🙏

**Comment by ousc** [comment]
Or is there a way to run two independent ng doc in the same angular program
